# Chapter 4 - Discounted Cash Flow Valuation

**Selected Exercises**

Corporate Finance 9th International Edition

Ross, Westerfield, Jaffe

In [1]:
import warnings
import numpy as np
import pandas as pd
import numpy_financial as npf
from helper_functions import *
from IPython.display import display, HTML

In [ ]:
"""
Chapter's Formulas:

Present Value of Investment =  C / ( 1 + r ) ^ T 
                             
Net Present Value of Investment = -Cost + PV of Future Cash Flows

Future Value of Investment = C0 * ( 1 + r ) ^ T

Effective Annual Rate (EAR) = ( 1 + r / m ) ^ m - 1

Future Value With Compounding = C0 * ( 1 + r / m) ^ m*T

Future Value With Continous Compounding = C0 * e ^ r*T (where e = 2.718)

               C
Perpetuity =  ---
               r

                       C
Growing Perpetuity = -----
                     r - g

                               [             1          ]
Present Value of Annuity = C * [ 1  -  --------------   ]
                               [        ( 1 + r) ^ T    ]
                               [ --------------------   ]
                               [           r            ]

                                           
                                       [       ( 1 + g ) ^ T  ]
Present Value of Growing Annuity = C * [ 1  -  ( ----- )      ]
                                       [       ( 1 + r )      ]
                                       [ -------------------- ] 
                                       [         r - g        ]
"""

## 1. Simple vs Compound Interest

First City Bank pays 9 percent simple interest on its savings account balances, whereas Second City Bank pays 9 percent interest compounded annually.  If you made a a 5,000 deposit in each bank, how much more money would you earn from your Second City Bank account at the end of 10 years?

In [3]:
interest_rate = .09
deposit = 5000
periods = 10

First_Bank = future_value_simple(deposit, interest_rate, periods)
Second_Bank = future_value_compounded(deposit, interest_rate, periods)

df = pd.DataFrame([[First_Bank, Second_Bank]],
                 columns = ("First Bank","Second Bank"))

df = df.style.set_properties(**{'text-align': 'center'})
df

,First Bank,Second Bank
0,"9,500.00","11,836.82"


## 2. Calculating Future Values

Compute the future value of 1,000 compounded annually for: 

* a. 10 years at 6 percent
* b. 10 years at 9 percent
* c. 20 years at 6 percent

In [4]:
#Read data
df = pd.DataFrame([[1000, 10, .06],
                    [1000, 10, .09],
                    [1000, 20, .06]],
                    columns=("Amount", "Periods", "Rate"))

#Apply Future Value function
df['Answer'] = df.apply(lambda row: future_value_compounded(row['Amount'], row['Rate'], row['Periods']), axis=1)

#Format Data Frame Style
df['Amount'] = df['Amount'].map('{:,.0f}'.format)
df['Rate'] = df['Rate'].map('{:.2%}'.format)
df = df.style.set_properties(**{'text-align': 'center'})
df

,Amount,Periods,Rate,Answer
0,"1,000",10,6.00%,"1,790.85"
1,"1,000",10,9.00%,"2,367.36"
2,"1,000",20,6.00%,"3,207.14"


## 3. Calculating Present Values

For each of the following compute the present value:

In [ ]:
"""
Years    Rate    Future Value
  6       7%        15,451
  9       15%       51,557
  18      11%      886,073
  23      18%      550,164
"""

In [5]:
#Read data
df = pd.DataFrame([[15541, 6, .07], 
                  [51557, 9, .15],
                  [886076, 18, .11],
                   [550164, 23, .18]],
                 columns = ("Future_Value", "Years", "Rate"))

#Apply Present Value function
df['Answer'] = df.apply(lambda row: present_value_amount(row['Future_Value'], row['Rate'], row['Years']), axis=1)

#Format Data Frame Style
df['Future_Value'] = df['Future_Value'].map('{:,.0f}'.format)
df['Answer'] = df['Answer'].map('{:,.0f}'.format)
df['Rate'] = df['Rate'].map('{:.2%}'.format)
df = df.style.set_properties(**{'text-align': 'center'})
df

,Future_Value,Years,Rate,Answer
0,"15,541",6,7.00%,"10,356"
1,"51,557",9,15.00%,"14,656"
2,"886,076",18,11.00%,"135,412"
3,"550,164",23,18.00%,"12,224"


## 4. Calculating Interest Rates

Solve for the unknown interest rate in each of the following:

In [ ]:
"""
Years    Present_Value    Future Value
  2            242           307
  9            410           896
 15           51,700       162,181
 30           18,750       483,500
"""

In [6]:
#Read data
df = pd.DataFrame([[2, -242, 307],
                  [9, -410, 896],
                  [15, -51700, 162181],
                  [30, -18750, 483500]],
                 columns = ("Years", "Present_Value", "Future_Value"))

# Solve for the interest rate
df["Answer"] = df.apply(lambda row: npf.rate(row["Years"], 0, \
                row["Present_Value"], row["Future_Value"]), axis=1)

#Format Data Frame Style
df['Future_Value'] = df['Future_Value'].map('{:,.0f}'.format)
df['Present_Value'] = df['Present_Value'].map('{:,.0f}'.format)
df['Answer'] = df['Answer'].map('{:.2%}'.format)
df = df.style.set_properties(**{'text-align': 'center'})
df

,Years,Present_Value,Future_Value,Answer
0,2,-242,307,12.63%
1,9,-410,896,9.07%
2,15,"-51,700","162,181",7.92%
3,30,"-18,750","483,500",11.44%


## 5. Calculating the Number of Periods

Solve for the unknown number of years in each of the following:

In [ ]:
"""
Present Value    Interest Rate    Future Value     
     625              6%              1,284
     810              13%             4,341
    18,400            32%            402,662
    21,500            16%            173,439
"""

In [7]:
warnings.filterwarnings("ignore", category=RuntimeWarning)

# Read Data
df = pd.DataFrame([[-625, .06, 1284],
                  [-810, .13, 4341],
                  [-18400, .32, 402662],
                  [-21500, .16, 173439]],
                 columns = ("Present_Value", "Rate", "Future_Value"))

#Solve for the years
df["Answer"] = df.apply(lambda row: npf.nper(row["Rate"], 0, \
                row["Present_Value"], row["Future_Value"]), axis=1)

#Format Data Frame Style
df['Future_Value'] = df['Future_Value'].map('{:,.0f}'.format)
df['Present_Value'] = df['Present_Value'].map('{:,.0f}'.format)
df['Answer'] = df['Answer'].map('{:,.2f}'.format)
df['Rate'] = df['Rate'].map('{:.2%}'.format)
df = df.style.set_properties(**{'text-align': 'center'})
df

,Present_Value,Rate,Future_Value,Answer
0,-625,6.00%,"1,284",12.36
1,-810,13.00%,"4,341",13.74
2,"-18,400",32.00%,"402,662",11.11
3,"-21,500",16.00%,"173,439",14.07


## 6. Calculating the Number of Periods

At 9 percent interest, how long does it take to double your money? To quadruple it?

In [20]:
"""
Use the Future Value formula and solve for T:
FV = PV * ( 1 + r ) ^ T
T = ln( PV / FV ) / ln( 1 + r )
"""
r = .09
investment = 1
doubled = investment * 2
doubled = np.log(doubled) / np.log(1 + r)

quadrupled = investment * 4
quadrupled = np.log(quadrupled) / np.log(1 + r)

print(f'To double it: {doubled:.0f}')
print(f'To quadruple it: {quadrupled:.0f}')


To double it: 8
To quadruple it: 16


## 7. Calculating Present Values

Imprudential Inc., has an unfunded pension liability of 750 million that must be paid in 20 years.  To assess the value of the firm's stock, financial analysts want to discount this liability back to the present.  If the relevant discount rate is 8.2 percent, what is the present value of this liability.

In [8]:
#Given
liability = 750000000
years = 20
rate = 8.2

#The present value function was defined at the beginning of the file.
present_value = present_value_amount(liability, rate / 100, years)

answer = f'The present value of the liability is ${present_value:,.2f}'
answer

'The present value of the liability is $155,065,808.54'

## 8. Calculating Rates of Return

Although applealing to more refined tastes, art as a collectible has not always performed so profitably.  During 2003, Sotheby's sold the Edgar Degas bronze sculpture *Petite Danseuse de Quartorze Ans* at auction for a price of 10,311,500.  Unfortunately for the previous owner, he had purchased it in 1999 at a price of 12,377,500. What was his annual rate of return on this sculpture?

In [9]:
# Given
years = 2003 - 1999
present_value = -12377500
future_value = 10311500

#We need to solve for rate:
rate = npf.rate(years, 0, present_value, future_value) * 100

answer = f'His annual rate of return on the sale of the sculpture was {rate:.2f}%'
answer

'His annual rate of return on the sale of the sculpture was -4.46%'

## 9. Perpetuities

An investor purchasing a British consol is entitled to receive annual payments from the British government forever.  What is the price of a consol that pays 120 annually if the next payment occurs one year from today?  The market interest is rate is 5.7 percent.

In [10]:
#Given
payments = 120
rate = 5.7

#We use the perpetuity function defined at the beginning
price = perpetuity(payments, rate / 100)

answer = f'The price of the console is {price:,.2f}'
answer


'The price of the console is 2,105.26'

## 10. Continuous Compounding

Compute the future value fo 1,900 continuously compounded for:

* a. 5 years at a stated annual interest rate of 12 percent.
* b. 3 years at a stated annual interest rate of 10 percent.
* c. 10 years at a stated annual interest rate of 5 percent.
* d. 8 years at a stated annual interest rate of 7 percent.

In [11]:
# Read data
df = pd.DataFrame([[1900, 5, .12],
                   [1900, 3, .10],
                   [1900, 10, .05],
                   [1900, 8, .07]],
                 columns = ("Present_Value", "Years", "Rate"))

# Compute values using continuous compounding function from helper_functions.
df['Answer'] = df.apply(lambda row: continuousFV(row['Present_Value'], \
                    row['Rate'], row['Years']), axis=1)

#Format Data Frame Style
df['Answer'] = df['Answer'].map('{:,.0f}'.format)
df['Present_Value'] = df['Present_Value'].map('{:,.0f}'.format)
df['Rate'] = df['Rate'].map('{:.2%}'.format)
df = df.style.set_properties(**{'text-align': 'center'})
df

,Present_Value,Years,Rate,Answer
0,"1,900",5,12.00%,"3,462"
1,"1,900",3,10.00%,"2,565"
2,"1,900",10,5.00%,"3,132"
3,"1,900",8,7.00%,"3,326"


## 11. Present Value and Multiple Cash Flows

Conoly Co. has identified an investment project with the following cash flows.  If the discount rate is 10 percent, what is the present value of these cash flows?  What is the value at 18 percent?  At 24 percent?

In [ ]:
"""
Year     Cash Flow
1          $1,200
2             730
3             965
4           1,590
"""

In [12]:
# For this exercise we must find the present value of each cash flow and then
# add them.  Extra work is required to get present values at different rates.

#Given
cash_flows = [1200, 730, 965, 1590]
rates = [.10, .18, .24]

year = 1
results = []

for rate in rates:
    for cash in cash_flows:
        pv = present_value_amount(cash, rate, year)
        results.append(pv)
        year += 1
    year = 1
    
present_values = {"10%" : results[0:4],
                "18%" : results[4:8],
                "24%" : results[8:]}

df = pd.DataFrame(present_values)
df.loc['Answer'] = df.sum()

#Format DataFrame style
df['10%'] = df['10%'].map('{:,.2f}'.format)
df['18%'] = df['18%'].map('{:,.2f}'.format)
df['24%'] = df['24%'].map('{:,.2f}'.format)
df


,10%,18%,24%
0,"1,090.91","1,016.95",967.74
1,603.31,524.27,474.77
2,725.02,587.33,506.13
3,"1,085.99",820.10,672.53
Answer,"3,505.23","2,948.66","2,621.17"


## 12. Present Value and Multiple Cash Flows

Investment X offers to pay you 5,500 per year for nine years, whereas Investment Y offers to pay you 8,000 per year for five years.  Which of these cash flows streams has the higher present value if the discount rate is 5 percent? If the discount rae is 22 percent?

In [13]:
rates = [.05, .22]

data = {
    "Rates" : rates,
    "Investment_X" : [annuityPVFactor(5500, x, 9) for x in rates],
    "Investment_Y" : [annuityPVFactor(8000, x, 5) for x in rates]
}

df = pd.DataFrame(data)

df["Higher_PV"] = df.apply(lambda row: higher_pv(row["Investment_X"], \
                row["Investment_Y"]), axis=1)

#Format DataFrame Style
df['Investment_X'] = df['Investment_X'].map('{:,.2f}'.format)
df['Investment_Y'] = df['Investment_Y'].map('{:,.2f}'.format)
df['Rates'] = df['Rates'].map('{:.0%}'.format)
df = df.style.set_properties(**{'text-align': 'center'})

df


,Rates,Investment_X,Investment_Y,Higher_PV
0,5%,"39,093.02","34,635.81",X
1,22%,"20,824.57","22,909.12",Y


**Analysis**

X has a greater PV at 5% but a lower PV at 22%, the reason is that X has greater total cash flows.  At a lower rate total cash flow is more important since the cost of waiting (the interest rate) is not as great.

On the other hand, Y is more valuable at a higher interest rate because it has larger cash flows.  At a higher rate these larger cash flows early are more important since the cost of waiting (the interest rate) is so much greater.

## 13. Calculating Annuity Present Value

An investment offers 4,300 per year for 15 years, with the first payment occurring one year from now.  If the required return is 9 percent, what is the value of the investment?  What would the value be if the payments occurred ofr 40 years? For 75 years? Forever?

In [14]:
payment = 4300
rate = .09

annuities = {
    "15y" : annuityPVFactor(payment, rate, 15),
    "40y" : annuityPVFactor(payment, rate, 40),
    "75y" : annuityPVFactor(payment, rate, 75),
    "Forever" : perpetuity(payment, rate) 
}

df = pd.DataFrame(annuities, index=['Answer'])
df['15y'] = df['15y'].map('{:,.2f}'.format)
df['40y'] = df['40y'].map('{:,.2f}'.format)
df['75y'] = df['75y'].map('{:,.2f}'.format)
df['Forever'] = df['Forever'].map('{:,.2f}'.format)
df = df.style.set_properties(**{'text-align': 'center'})

df

,15y,40y,75y,Forever
Answer,"34,660.96","46,256.65","47,703.26","47,777.78"


## Exercise 14 - Calculating Perpetuity Values 

The Perpetual Life Insurance Co. is trying to sell you an investment policy that will pay you and you heirs 20,000 per year forever.  

* a. If the required return on this investment is 6.5 percent, how much will you pay for the policy?
* b. Suppose the Perpetual Life Insurance Co. told you the policy costs 340,000.  At what interest rate would this be a fair deal?

In [15]:
#given
payment = 20000
offer = 340000
rate = 6.5

# Lets first find the price of the perpetuity
a = perpetuity(payment, rate/100)

# Lets now find a fair rate for the price been asked.
b = payment / offer

answer = f'Answers:\na. I would pay {a:,.2f} for the perpetuity.\nb. \
{b*100:.2f}% would be a fair deal.'

print(answer)

Answers:
a. I would pay 307,692.31 for the perpetuity.
b. 5.88% would be a fair deal.


## Exercise 15. Calculating EAR - Find the EAR in each of the following cases:

In [ ]:
"""
Stated Rate(APR)    Compounded
--------------------------------
8%                   Quarterly
18%                  Monthly
12%                  Daily
14%                  Infinite
---------------------------------
"""

In [16]:
# Read data.
df = pd.DataFrame([[0.08, 4],
                  [0.18, 12],
                  [0.12, 365]],
                 columns=("APR", "Compounding"))

# Apply the EAR function.
df["EAR"] = df.apply(lambda row: effectiveRate(row["APR"], row["Compounding"]), axis=1)

#Format DataFrame Style.
df['APR'] = df['APR'].map('{:.2%}'.format)
df['EAR'] = df['EAR'].map('{:.2%}'.format)
df = df.style.set_properties(**{'text-align': 'center'})
df


,APR,Compounding,EAR
0,8.00%,4,8.24%
1,18.00%,12,19.56%
2,12.00%,365,12.75%


## Exercise 16 - Calculating APR

Find the APR, or stated rate, in each of the following cases:

In [ ]:
"""
   Effective 
Annual Rate(EAR)     Compounded
----------------------------------
    10.3 %           Semiannually
     9.4               Monthly
     7.2                Weekly
    15.9               Infinite
----------------------------------
"""

Lets first define a function that would allow us to convert from EAR to APR.  As you may recall, we find the EAR as follow:

In [ ]:
"""
Effective Annual Rate (EAR) = ( 1 + r / m ) ^ m - 1
"""

We rearrange the equation to solve for r (APR). 

In [ ]:
def aprRate(ear, m):
    apr = m * ((1 + ear)**(1/m) - 1)
    return apr

Ok, perfect, now we create our table as shown below.

In [10]:
df = pd.DataFrame([[0.103, 2],
                  [0.094, 12],
                  [0.072, 52]],
                 columns=("EAR", "Compounding"))

df["APR"] = df.apply(lambda row: aprRate(row["EAR"], row["Compounding"]), axis=1)

#Format DataFrame Style.
df['EAR'] = df['EAR'].map('{:.2%}'.format)
df['APR'] = df['APR'].map('{:.2%}'.format)
df = df.style.set_properties(**{'text-align': 'center'})
df

,EAR,Compounding,APR
0,10.30%,2,10.05%
1,9.40%,12,9.02%
2,7.20%,52,6.96%


Solving APR indefinately is a bit different, since we need to use the `Continuous Compounding` formula.

EAR = e^q - 1

Now we rearrange and solve for APR:

In [14]:
ear = .1590
apr = np.log(1 + ear) * 100
print(f"{apr:.2f}%")

14.76%


## Exercise 17 - Calculating EAR

First National Bank charges 10.1 percent compounded monthly on its business loans.  First United Bank charges 10.4 percent compounded semiannually.  As a potential borrower, to which bank would you go for a new loan?

In [29]:
# Lets create a simple data structure.
struct = {"Interest": [.101, .104],
          "Compounding": [12, 2]
         }

# Lets create a table
df = pd.DataFrame(struct)
df.index = ["First_National_Bank", "First_United_Bank"]

# Create an EAR column and apply the helper function
df["EAR"] = df.apply(lambda row: effectiveRate(row["Interest"], row["Compounding"]), axis=1)

#Format DataFrame Style.
df['Interest'] = df['Interest'].map('{:.2%}'.format)
df['EAR'] = df['EAR'].map('{:.2%}'.format)
df = df.style.set_properties(**{'text-align': 'center'})
df

,Interest,Compounding,EAR
First_National_Bank,10.10%,12,10.58%
First_United_Bank,10.40%,2,10.67%


First National Bank has the lower APR, but having the lowest APR is not the most important factor. Compounding is very important and when taken into account they are almost tied.  Nevertheless, I would go with First National Bank for this one.

## 19. Calculating Number of Periods

One of your customers is delinquent on his accounts payable balance.  You've mutually agreed to a repayment schedule of `$600 per month`.  You will charge `.9 percent per month interest` on the overdue balance.  If the `current balance is 18,400`, how long will it take for the account to be paid off?

In [5]:
r = .009
pmt = 600
fv = 0
balance = 18400
months = npf.nper(pv=-balance, fv=fv, rate=r, pmt=600)
print(f"The loan will be paid-off in approximately: {months:.2f} months.")

The loan will be paid-off in approximately: 36.05 months.


## 20. Calculating EAR

Friendly's Quick Loans, Inc., offeres you "three for four or 1 knock on your door."  This means you get `3 dollars` today and repay `4 dollars` when you get your paycheck in `one week` (or else).  What's the effective annual return Friendly's earns on this lending business?  If you were brave enough to ask, what APR would Friendly's say you were paying?

1. Lets find the interest rate per week.

In [2]:
pv = -3
fv = 4
n = 1
pmt = 0

weekly_interest = npf.rate(pv=pv, fv=fv, nper=1, pmt=0)
print(f"The interest per week is: {weekly_interest*100:.2f}%")

The interest per week is: 33.33%


2. The APR is the interest per week times 52 weeks.

In [8]:
apr = weekly_interest * 52
print(f"APR: {apr*100:,.2f}%")

APR: 1,733.33%


3. To get the Effective Rate we use the EAR equation.

In [11]:
m = 52 # compounding
ear = (1 + (apr / m))**m - 1
print(f"The effective annual rate (EAR) is: {ear*100:,.2f}%")

The effective annual rate (EAR) is: 313,916,515.69%


## 21. Future Value 

What is the future value in `seven years` of `1,000` invested in an account with a stated annual interest rate of `8 percent`.

1. Compounded annually?
2. Compounded semiannually?
3. Compounded monthly?
4. Compounded continuously?

In [53]:
# Get Variables
n = 7
r = .08
e = 2.718
pmt = 0
pv = -1000
fv = []

# Create a basic data structure
struct = {"Compoundings" : ["Yearly", "Semiannual", "Monthly", "Continuous"],
          "Future_Values" : [] 
         }

# Compoundings
# 1 = Yearly, 2 = Semiannual, 12 = Monthly, e = continuous
compoundings = [1, 2, 12, e]

# Iterate over the compounding list and compute future values.
# Also populate Future_Values in struct.

for m in compoundings:
    if m == e:
        fvlue = np.abs(pv) * (e**(r*n))
        struct["Future_Values"].append(fvlue)
    else:
        comp_rate = r / m
        comp_perd = n * m
        fvlue = npf.fv(pv=pv, rate=comp_rate, nper=comp_perd, pmt=0)
        struct["Future_Values"].append(fvlue)

# Create and Display answer table
df = pd.DataFrame(struct)
df['Future_Values'] = df['Future_Values'].map('{:,.2f}'.format)
display(HTML(df.to_html(index=False)))

Compoundings,Future_Values
Yearly,"1,713.82"
Semiannual,"1,731.68"
Monthly,"1,747.42"
Continuous,"1,750.57"


## 23. Calculating Annuities

You are planning to save for retirement over the next 30 years.  To do this, you will invest 700 dollars a month in a stock account and 300 a month in a bond account.  The return of the stock account is expected to be 10 percent, and the bond account will pay 6 percent.  When you retire, you will combine your money into an account with an 8 percent return. How much can you withdraw each month from your account assumming a 25 year withdrawal period?

Since the investing period ends at the same time as the withdrawals begins, the present value of the retirement payments is equal to the future value of the retirement savings. So we need to get the FV of the stock account and the fv of the bond account and add them together.

In [16]:
save_period = 30
stock_ret = .10
bond_ret = .06
cmp = 12 # Assuming monthly compounding
stock_pmts = 700
bond_pmts = 300

struct = {
    "Stock_Account" : -npf.fv(pv=0, nper=save_period*cmp, rate=stock_ret/cmp, pmt=stock_pmts),
    "Bond_Account" : -npf.fv(pv=0, nper=save_period*cmp, rate=bond_ret/cmp, pmt=bond_pmts)
    }

savings = pd.DataFrame(struct, index=["Future Value"])
savings["Total_Savings"] = savings["Stock_Account"] + savings["Bond_Account"]
savings["Stock_Account"] = savings["Stock_Account"].map('{:,.2f}'.format)
savings["Bond_Account"] = savings["Bond_Account"].map('{:,.2f}'.format)
savings["Total_Savings"] = savings["Total_Savings"].map('{:,.2f}'.format)
savings


,Stock_Account,Bond_Account,Total_Savings
Future Value,"1,582,341.55","301,354.51","1,883,696.06"


After the end of our saving period we have accumulated `1,883,696 dollars` which are ready for funding our retirement.  We find the periodic distribution by using the `pmt` function.

In [30]:
r = .08 # combined funds at an 8 percent account
n = 25 # Estimated Withdrawal period
cmp = 12 # Assuming monthly compounding
pv = struct["Stock_Account"] + struct["Bond_Account"]
distribution = -npf.pmt(rate=r/cmp, nper=n*cmp, pv=pv, fv=0, when='end')
print(f"Expected Monthly Distribution: ${distribution:,.2f}")

Expected Monthly Distribution: $14,538.67


## 24. Calculating Rates of Return

Suppose an investment offers to quadruple your money in 12 months (don't believe it).  What rate of return per quarter are you being offered?

In [32]:
fv = 4
pv = 1
n = 12
m = 3
rate = npf.rate(fv=fv, pv=-pv, nper=n/m, pmt=0)
print(f"The rate of return per quarter is: {rate*100:.2f}%")

The rate of return per quarter is: 41.42%


## 25. Calculating Rates of Return

You're trying to choose between two different investments, both of which have up-front costs of `75,000` dollars.  Investment G returns `135,000` dollars in `six years`.  Investment H returns `195,000` dollars in `10 years`.  Which of these investments has the higher return?

In [22]:
cost = 75000

struct = {
    "Investment_G" : [135000, 6],
    "Investment_H" : [195000, 10],   
}

for k, v in struct.items():
    r = npf.rate(pv=-cost, fv=v[0], nper=v[1], pmt=0)
    v.append(r)

df = pd.DataFrame(struct)
df.index = ["Returns", "Years", "Rates"]
df_t = df.T
df_t["Returns"] = df_t["Returns"].map('{:,.2f}'.format)
df_t["Years"] = df_t["Years"].astype('float').map('{:.0f}'.format)
df_t["Rates"] = df_t["Rates"].map('{:.2%}'.format)
df_t

,Returns,Years,Rates
Investment_G,"135,000.00",6,10.29%
Investment_H,"195,000.00",10,10.03%


Investment G has the higher return of 10.29% in comparison with Investment H which only has 10.03%

## 26. Growing Perpetuities

Mark Weinstein has been working on an advanced technology in laser eye surgery.  His technology will be available in the near term.  He anticipates his first annual cash flow from the technology to be `215,000`, received `two years from today`.  Subsequent cash flows will `grow` at `4 percent` in `perpetuity`.  What is the `present value` of the technology if the discount rate is `10 percent`?

To solve this problem, we first need to get the present value of the growing perpetuity at date 1, and then, the present value of a lump sum at date 0:

In [ ]:
"""
Date                   0               1                2               3              Forever
                       |---------------|----------------|---------------|----------------|
Cash Flow                                          $215,000        215,000(1 + g) 
                       _________________________________|
                       |               |
Present Value     $3,257,575       $3,583,333
"""

From the exercise's text we get the following variables:

In [35]:
r = .10 # Discount rate
g = .04 # Growth rate
c_flow = 215000 # first cash flow

Lets define a helper function for the growing perpetuity equation:

In [27]:
def growing_perpetuity(c_flow, r, g):
    g_perpetuity = (c_flow / (r - g))
    return g_perpetuity

The present value of the growing perpetuty at date 1 is:

In [30]:
perpetuity = growing_perpetuity(c_flow, r, g)
print(f"${perpetuity:,.2f}")

$3,583,333.33


So the value of the perpetuity is 3.6 million dollars at date one, but we must find the present value of the perpetuity at date 0:

In [36]:
n = 1
pv = -npf.pv(fv=perpetuity, rate=r, nper=n, pmt=0)
print(f"Present Value at date 0: ${pv:,.2f}")

Present Value at date 0: $3,257,575.76


## 27. Perpetuities

A prestigious investment bank designed a new security that pays a quarterly dividend of 5 dollars in perpetuity.  The first dividend occurs one quarter from today.  What is the price of the security if the stated annual interest rate is 7 percent, compounded quarterly?

In [4]:
r = .07
m = 4
div = 5

price = div / (r / m)
print(f"The price of the perpetuity is: ${price:.2f}")

The price of the perpetuity is: $285.71


## 28. Annuity Present Values

What is the present value of an `annuity` of `5,000` per year, with the `first cash flow` received `three years` from today and the last one received `25 years from today`?  Use a discount rate of `8 percent`.

To solve this exercise, we need to first find the present value of the annuity at date 2.  Then we find the present value of a lump sum at date 0.

In [ ]:
"""
Date                   0           1          2          3    ...   25 
                       |-----------|----------|----------|----------|
Cash Flows                                 $5,000     $5,000  ... $5,000
                        ______________________|
                       |                      |
Present Value       $44,457.56             $51,855.29 
"""

In [9]:
r = .08
c_flow = 5000
n1 = 25 - 2

annuity_pv = -npf.pv(fv=0, rate=r, nper=n1, pmt=c_flow)
print(f"The pv of the annuity at date 2 is: ${annuity_pv:,.2f}")

The pv of the annuity at date 2 is: $51,855.29


Now that we have the present value of the annuity at date 2, we can compute the present value as of today.

In [12]:
n2 = 2
pv = -npf.pv(fv=annuity_pv, rate=r, nper=n2, pmt=0)
print(f"The present value as of the annuity today is: ${pv:,.2f}")

The present value as of the annuity today is: $44,457.56


## 29. Annuity Present Value

What is the value `today` of a `15-year annuity` that pays `750 a year`?  The annuity's `first payment` occurs `six years from today`.  The annual interest rate is `12 percent` for `years 1 through 5`, and `15 percent thereafter`.

To find the present value of the 15-year annuity today, we need to find the present value of the annuity at date 5 and then find the present value of the lump sum at date 0.

In [ ]:
"""
Date              0      1     2     3     4     5     6     7 ... 20
                  |------|-----|-----|-----|-----|-----|-----|-----|
Cash Flow                                             750   750   750
                  _____________________________________|    
                 |                               |
Present Value   2,488                           4,385.53
"""

In [17]:
r1 = .12 # Interest rate from year 1 through 5
r2 = .15 # Interest rate thereafter
n = 15 # First 5 periods
pmt = 750 # Per year

1. Find the present value of the annuity at date 5:

In [20]:
annuity_pv = -npf.pv(fv=0, rate=r2, nper=n, pmt=pmt)
print(f"Present value at date 5: ${annuity_pv:,.2f}")

Present value at date 5: $4,385.53


2. Present value at date 0:

In [24]:
pv = -npf.pv(fv=annuity_pv, rate=r1, nper=5, pmt=0)
print(f"Present value at date 0: ${pv:,.2f}")

Present value at date 0: $2,488.47


## 30. Balloon Payments 

Audrey Sandborn has just arranged to purchase a `450,000` vacation home in the Bahamas with a `20 percent down payment`.  The mortgage has a `7.5 percent` stated annual interest rate, `compounded monthly`, and calls for equal monthly payments over the next `30 years`.  Her first payment will be due one month from now.  However, the mortgage has an `eight-year balloon payment`, meaning that the balance of the loan must be paid off at the end of year 8.  There were no other transaction costs or finance charges.  How much will Audrey's balloon payment be in eight years?

1. **From the exercise's text we can extract the following data:**

In [18]:
m = 12                           # Monthly compounding
interest = .075 / m              # Monthly interest rate
loan = 450000                    # Mortgage
down_payment = loan * .20        # 20% Down Payment
balance = loan - down_payment    # Adjusted mortgage after 20% payment
term = 30 * m                    # Monthly periods over 30 Years
balloon_terms = 8 * 12            # Eighth Year Balloon Agreement

2. **Since the exercise does not provide the monthly payment, we must calculate what the payment is:**

In [9]:
pmt = npf.pmt(pv=-balance, fv=0, nper=term, rate=apr)

np.float64(2517.172230789996)

3. **Lets see what data we have so far:**

In [23]:
print(f"Balance: ${balance:,.2f}\nPayment: ${pmt:,.2f}\nInterest: {interest*100:.4f}%")
print(f"Monthly periods: {term}\nBalloon Terms: {balloon_term}")

Balance: $360,000.00
Payment: $2,517.17
Interest: 0.6250%
Monthly periods: 360
Balloon Terms: 96


4. **Using the easy way, we realize that after the balloon terms of 96 months we have a loan of:**

In [22]:
remaining_periods = term - balloon_terms
print(f"{remaining_periods} monthly periods.")

264 monthly periods.


5. **So, the balloon payment is equal to the present value of the remaining monthly payments on the loan.**

In [27]:
balloon_pmt = -npf.pv(fv=0, rate=interest, nper=remaining_periods, pmt=pmt)
print(f"Balloon Payment: ${balloon_pmt:,.2f}")

Balloon Payment: $325,001.73


## 31. Calculating Interest Expense

You receive a credit card application from Shady Banks savings and Loan offering an introductory rate of `2.40 percent per year`, `compounded monthly for the first six months, increasing thereafter to 18 percent compounded monthly`.  Assuming you transfer the `6,000 balance` from your existing credit card and make `no subsequent payments`, how much interest will you owe at the end of the first year?

In [ ]:
m = 12                    # Monthly compounding
n = 6                     # Six months each
balance = 6000            # Credit Card Transfer
intro_int = .0240 / m     # Introductory interest rate per month             
final_int = .18 / m       # 18% Thereafter

1. **Lets first compute the interest of the first six months**

In [36]:
first_six_interest = npf.fv(pv=-balance, rate=intro_int, nper=n, pmt=0)
print(f"Introductory Interest: ${first_six_interest:,.2f}")

Introductory Interest: $6,072.36


2. **Now lets calculate the interest for the remaining six month of the year.  Note that our balance has increased due to the unpaid interest calculated above.**

In [37]:
thereafter_interest = npf.fv(pv=-first_six_interest, rate=final_int, nper=n, pmt=0)
print(f"Thereafter Interest: {thereafter_interest:,.2f}")

Thereafter Interest: 6,639.78


3. **Ok fine, up to this point we have the total balance plus interest outstanding.  To get the interest expense we subtract the original balance of 6,000:**

In [39]:
interest_expense = thereafter_interest - balance
print(f"Interest Expense: ${interest_expense:,.2f}")

Interest Expense: $639.78


## 32. Perpetuities

Barret Pharmaceuticals is considering a drug project that costs a `150,000 today` and is expected to generate end-of-year annual `cash flows of 13,000 forever`.  At what discount rate would Barret be indifferent between accepting or rejecting the project?

**1. The rate that would make Barret indifferent between accepting or rejecting the project is the rate that makes the present value of the cash flows equal to the cost today.  Since the cash flows are a perpetuity, we can use the present value of perpetuity equation:**

In [42]:
cost = 150000
c_flows = 13000

# 150000 = 13000 / r
r = c_flows / cost
print(f"The rate that would make Barren indifferent is: {r*100:.2f}%")

The rate that would make Barren indifferent is: 8.67%


## 33. Growing Annuity

Southern California Publishing Company is trying to decide whether to revise its popular textbook, *Financial Psychoanalysis Made Simple*.  The company has estimated that the revision will `cost 65,000 dollars`.  `Cash flows` from increased sales will be `18,000 the first year`.  These cash flows will `increase by 4 percent per year`.  The book will go out of print five years from now.  Assume that the initial `cost is paid now and revenues are received at the end of each year`.  If the company requires an `11 percent` return for such an investment, should it undertake the revision? 

In [44]:
r = .11
g = .04
n = 5
cost = 65000
c_flows = 18000

In [ ]:
"""
                                       [       ( 1 + g ) ^ T  ]
Present Value of Growing Annuity = C * [ 1  -  ( ----- )      ]
                                       [       ( 1 + r )      ]
                                       [ -------------------- ] 
                                       [         r - g        ]
"""

**Lets solve this problem by getting the present value of the growing annuity and then comparing it to the cost. For this purposes, we define a helper function for the pv of growing annuity equation.**

In [49]:
def growing_annuity(c, r, g, n):
    fact = (1 - ((1 + g) / (1 + r))**n) / (r - g)
    return c * fact

In [47]:
annuity_pv = growing_annuity(c_flows, r, g, n)
print(f"Present Value of Annuity: ${annuity_pv:,.2f}")

Present Value of Annuity: $71,479.47


**Southern California Publishing Company should accept the project because the present value of the growing annuity is greater that the cost.**

## 34. Growing Annuity

Your job pays you only once a year for all the work you did over the previous 12 months.  Today, December 31, you just received your salary of `60,000 dollars`, and you plan to spend all of it.  However, you want to start savings for retirement beginning next year.  You have decided that one year from today you will begin depositing `5 percent of your annual salary` in an account that will earn 9` percent` per year.  Your `salary will increase at 4 percent per year` throughout your career.  How much money will you have on the date of your retirement `40 years from today`?

**To solve this problem we must find the present value at date 0 of the growing perpetuity and then get the future value of the lump sum to find the amount accumulated at date 40.**

In [63]:
interest = .09
growth_rate = .04
current_salary = 60000
savings_ratio = .05
annual_deposit = (current_salary * (1 + growth_rate)) * savings_ratio 
savings_period = 40

**We can use the function defined at exercise 33 for the growing annuity equation.**

In [68]:
pv = growing_annuity(annual_deposit, interest, growth_rate, savings_period)
print(f"Present Value of Growing Annuity at Date 0: ${pv:,.2f}")

Present Value of Growing Annuity at Date 0: $52,861.98


**The amount accumulated over the savings period amounts to:**

In [67]:
fv = npf.fv(pv=-pv, rate=interest, nper=savings_period, pmt=0)
print(f"${fv:,.2f}")

$1,660,364.12


## 35. Present Values and Interest Rates

What is the relationship between the value of an annuity and the level of interest rates?  Suppose you just bought a 12-year annuity of 7,500 per year at the current interest rate of 10 percent per year.  What happends to the value of your investment if interest rates suddenly drop to 5 percent?  What if interest rates suddenly rise to 15 percent?

In [74]:
n = 12
annuity = 7500
rates = [.10, .05, .15]

for rate in rates:
    pv = -npf.pv(fv=0, nper=n, pmt=annuity, rate=rate)
    print(f"Investment at: {rate*100:.2f}% = ${pv:,.2f}")

Investment at: 10.00% = $51,102.69
Investment at: 5.00% = $66,474.39
Investment at: 15.00% = $40,654.64


**Relationship**

There is an inverse relationship, the present value falls as interest rates increases, and rises as interest rates decreases.  Why? because that lump sum could be generating higher revenues from higher interest rates, but it is not.  On the other hand, but in the same line, the future value rises as interest go up, and fall when interest go down.

## 36. Calculating the Number of Payments

You're prepared to make monthly payments of 250 dollars, beginning ath the end of this month, into an account that pays 10 percent interest compounded monthly.  How many payments will you have made when your account balance reaches 30,000 dollars?


In [90]:
cmp = 12          # Monthly compounding
rate = .10 / cmp  # Monthly rate
pmt = 250         # Monthly payment
f_value = 30000   # The future value

# Get Number of payments
payments = -npf.nper(pmt=pmt, rate=r, fv=0, pv=fv)
print(f"There will be {payments:.2f} payments when the account reaches ${f_value:,.2f}")

There will be 83.52 payments when the account reaches $30,000.00


## 37. Calculating Annuity Present Values

You want to borrow 80,000 from your local bank to buy a new sailboat.  You can afford to make monthly payments of 1,650 but no more.  Assuming monthly compounding, what is the highest APR you can afford on a 60-month loan?

In [9]:
n = 60
loan = 80000
pmt = 1650
rate = npf.rate(pv=-loan, nper=n, pmt=pmt, fv=0)
print(f"{rate:.2f}%")

0.01%


0.01 is the monthly interest rate, to get the APR: 

In [11]:
apr = rate * 12 * 100
print(f"The highest APR that can be afforded is: {apr:.2f}%")

The highest APR that can be afforded is: 8.72%


## 38. Calculating Loan Payments

You need a `30-year`, fixed-rate mortgage to buy a new home for `250,000`.  Your mortgage bank will lend you the money at a `6.8 percent APR` for this `360-month loan`.  However, you can only afford `monthly payments of 1,200`, so you offer to pay off any remainning loan balance `at the end of the loan` in the form of a `single balloon payment`.  How large will this balloon payment have to be for you to keep your monthly payments at 1,200?

In [12]:
r = .068 / 12
n = 360
pmt = 1200
loan = 250000

**The amount of principal paid on a loan is the present value of the periodic payments.**

In [20]:
pv = -npf.pv(fv=0, rate=r, nper=n, pmt=pmt)
print(f"Present Value of Payments: ${pv:,.2f}")

Present Value of Payments: $184,070.20


**The amount of principal still owed is the loan total minus the present value of payments**

In [23]:
owed = loan - pv
print(f"Amount owed: ${owed:,.2f}")

Amount owed: $65,929.80


**The problem is that this remaining principal will increase at the interest rate of the loan over the 30 year period.  So the balloon payment in 30 years, which is the future value of the remaining principal is:**

In [27]:
fv = npf.fv(pv=-owed, rate=r, nper=n, pmt=0)
print(f"Balloon payment: {fv:,.2f}")

Balloon payment: 504,129.05


## 39. Present and Future Values

The `present value` of the following cash flow stream is `6,453` when discounted at `10 percent annually`.  What is the value of the missing cash flow?

In [ ]:
"""

Year         Cash Flow
----         ---------
 1             1,200
 2               ?
 3             2,400
 4             2,600
----         --------

"""

**We know what the present value of the four cash flows is, If we could just find the present value of each cash flow, then we could subtract the present values of the three periods from the total present value to arrive at the present value of the missing cash flow.  To find the actual cash flow we will use the future value function.**

In [56]:
interest = .10                            
pvalue = 6453
cash_flows = [1200, 0, 2400, 2600]

sum_of_pvs = 0

for k, v in enumerate(cash_flows, start=1):
    if v == 0:
        pass
    else:
        pv = v / (1 + interest)**k
        sum_of_pvs += pv
        print(f"Present Values at period {k} = ${pv:,.2f}")

pv_of_missing_cf = pvalue - sum_of_pvs

print(f"{"-"*38}\nSum of known present values = ${sum_of_pvs:,.2f}")
print(f"Present Value of Missing Cash Flow = ${pv_of_missing_cf:,.2f}")

Present Values at period 1 = $1,090.91
Present Values at period 3 = $1,803.16
Present Values at period 4 = $1,775.83
--------------------------------------
Sum of known present values = $4,669.90
Present Value of Missing Cash Flow = $1,783.10


**So now that we found the present value of the missing cash flow, we can find the actual missing cash flow value using the future value function.**

In [57]:
missing_cash_flow = npf.fv(pv=-pv_of_missing_cf, nper=2, rate=interest, pmt=0)
print(f"Missing Cash Flow: ${missing_cash_flow:,.2f}")

Missing Cash Flow: $2,157.55


## 40. Calculating Present Values

You just won the TVM Lottery.  You will receive 1 million today plus another 10 annual payments that increase by 350,000 dollars per year.  Thus, in one year you receive 1.35 million.  In two years, you get 1.7 million, and so on. If the appropriate interest rate is 9 percent, what is the present value of your winnings?

In [76]:
rate = .09
periods =  10          
reg_annuity = 1000000
increment = 350000         

present_value = 1000000    # First payment today

for k, v in enumerate(range(periods), start=1):
    annuity = reg_annuity + increment
    present_value += (annuity / (1 + rate)**k)
    increment += 350000

print(f"Present Value of Winnings: ${present_value:,.2f}")

Present Value of Winnings: $18,194,308.69


## 41. EAR versus APR

You have just purchased a new warehouse.  To finance the purchase, you've arranged for a `30-year` mortgage for `80 percent` of the `2,600,000` purchase price.  The monthly `payment` on this loan will be `14,000`.  What is the APR on this loan? The EAR?

In [ ]:
"""
Effective Annual Rate (EAR) = (( 1 + r )^m) - 1
"""

In [91]:
m = 12                # Monthly compounding
periods = 30 * m      # 12 monthly payments per year
pmt = 14000           # Monthly payments

monthly_rate = npf.rate(nper=periods, pmt=pmt, pv=-loan, fv=0)
apr = monthly_rate
ear = (1 + apr)**m - 1

print(f"APR: {apr*100:,.4f}%")
print(f"EAR: {ear*100:,.2f}%")

APR: 0.5930%
EAR: 7.35%


## 42. Present Value and Break-Even Interest

Consider a firm with a contract to sell an asset for 135,000 dollars three years from now.  The asset costs 96,000 to produce today.  Given a relevant discount rate on this asset of 13 percent per year, will the firm make a profit on this asset?  At what rate does the firm just break even?

In [102]:
rate = .13
periods = 3
cost = 96000
price = 135000

profit = (price / (1 + rate)**periods) - cost
beven_rate = npf.rate(pv=-cost, fv=price, nper=periods, pmt=0)
print(f"Profit: ${profit:,.2f}")
print(f"Break Even Rate: {beven_rate*100:.2f}%")

Profit: $-2,438.23
Break Even Rate: 12.04%


## 43. Present Value and Multiple Cash Flows

What is the present value of `4,000 per year`, at a discount `rate of 7 percent`, if the first payment is received `9 years from now` and the last payment is received `25 years from now`?

**To solve this problem with need find the present value of the annuity at date 8 and then the present value of the lump sum at date 0**

In [ ]:
"""
Date             0           6          7          8          9   ...    25       
                 |-----------|----------|----------|----------|----------|
Cash Flow        |                                4000       4000 ...   4000
                 |                                 |                     |
Present Value   $22,729.14 <----------------  $39,052.89 <---------------|

"""

In [124]:
n1 = 8        # Discount periods for lump sum
n2 = 17       # Discount periods for annuity
rate = .07    # Discount rate
pmt = 4000    # Yearly Payment

annuity_pv = -npf.pv(rate=rate, nper=n2, pmt=pmt, fv=0)
pv = -npf.pv(rate=rate, nper=n1, pmt=0, fv=annuity_pv)

print(f"Present Value at Date 8: ${annuity_pv:,.2f}")
print(f"Present Value at Date 0: ${pv:,.2f}")

Present Value at Date 8: $39,052.89
Present Value at Date 0: $22,729.14
